CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+  f = File.open('testfile')
+  begin
+    # .. process
+  rescue
+    # .. handle error
+  ensure
+    f.close if f
+  end
+  ```
+
+* <a name="auto-release-resources"></a>
+Use versions of resource obtaining methods that do automatic
+resource cleanup when possible.
+<sup>[[link](#auto-release-resources)]</sup>
+
+  ```Ruby
+  # bad - you need to close the file descriptor explicitly
+  f = File.open('testfile')
+    # ...
+  f.close
+
+  # good - the file descriptor is closed automatically
+  File.open('testfile') do |f|
+    # ...
+  end
+  ```
+
+* <a name="standard-exceptions"></a>
+  Favor the use of exceptions for the standard library over introducing new
+  exception classes.
+<sup>[[link](#standard-exceptions)]</sup>
+
+## Collections
+
+* <a name="literal-array-hash"></a>
+  Prefer literal array and hash creation notation (unless you need to pass
+  parameters to their constructors, that is).
+<sup>[[link](#literal-array-hash)]</sup>
+
+  ```Ruby
+  # bad
+  arr = Array.new
+  hash = Hash.new
+
+  # good
+  arr = []
+  hash = {}
+  ```
+
+* <a name="percent-w"></a>
+  Prefer `%w` to the literal array syntax when you need an array of words
+  (non-empty strings without spaces and special characters in them).  Apply this
+  rule only to arrays with two or more elements.
+<sup>[[link](#percent-w)]</sup>
+
+  ```Ruby
+  # bad
+  STATES = ['draft', 'open', 'closed']
+
+  # good
+  STATES = %w(draft open closed)
+  ```
+
+* <a name="percent-i"></a>
+  Prefer `%i` to the literal array syntax when you need an array of symbols
+  (and you don't need to maintain Ruby 1.9 compatibility). Apply this rule only
+  to arrays with two or more elements.
+<sup>[[link](#percent-i)]</sup>
+
+  ```Ruby
+  # bad
+  STATES = [:draft, :open, :closed]
+
+  # good
+  STATES = %i(draft open closed)
+  ```
+
+* <a name="no-trailing-array-commas"></a>
+  Avoid comma after the last item of an `Array` or `Hash` literal, especially
+  when the items are not on separate lines.
+<sup>[[link](#no-trailing-array-commas)]</sup>
+
+  ```Ruby
+  # bad - easier to move/add/remove items, but still not preferred
+  VALUES = [
+             1001,
+             2020,
+             3333,
+           ]
+
+  # bad
+  VALUES = [1001, 2020, 3333, ]
+
+  # good
+  VALUES = [1001, 2020, 3333]
+  ```
+
+* <a name="no-gappy-arrays"></a>
+  Avoid the creation of huge gaps in arrays.
+<sup>[[link](#no-gappy-arrays)]</sup>
+
+  ```Ruby
+  arr = []
+  arr[100] = 1 # now you have an array with lots of nils
+  ```
+
+* <a name="first-and-last"></a>
+  When accessing the first or last element from an array, prefer `first` or
+  `last` over `[0]` or `[-1]`.
+<sup>[[link](#first-and-last)]</sup>
+
+* <a name="set-vs-array"></a>
+  Use `Set` instead of `Array` when dealing with unique elements. `Set`
+  implements a collection of unordered values with no duplicates. This is a
+  hybrid of `Array`'s intuitive inter-operation facilities and `Hash`'s fast
+  lookup.
+<sup>[[link](#set-vs-array)]</sup>
```

Comment:
I disagree with this in the case of very short lists for two reasons:

1. Ruby has no Set literal, so it is bulkier.
2. Array is faster for low cardinality sets.

I generally prefer Set, but I think code like: `[:a,:b].include?(x)` is much better than `Set.new([:a,b]).include?(x)`. Additionally, from a performance standpoint, the Set option is ludicrous in that it sets up a hash map, hashes each element, then has to hash the argument before doing the lookup. Constant factors count.